[evm] support arguments in evm-based unit tests

vgao1996 opened the pull request:
This allows arguments (`#[test(a = ...)]`) to be used in EVM-based unit tests. 

It should be noted that right now, the unit test framework itself only supports address arguments (passed to parameters of `signer` or `address` types), so this is all we get for EVM-based unit tests.

wrwg commented on language/evm/move-to-yul/src/native_functions.rs:
```diff
@@ -39,15 +39,12 @@ impl NativeFunctions {
-                        .get_full_name_str(),
-                    fun_id.inst,
-                ),
+            unimplemented!(
```
There is a simple rule which one can apply here: if a user input (or any external input) into a program can lead to an error situation, the program should never crash, but gracefully produce an error. Since a user can just declare any kind of native functions, we should better continue to produce an error.

vgao1996 replied:
> Since a user can just declare any kind of native functions, we should better continue to produce an error.

I see your point and I agree. Now I think this actually reveals a bigger problem: right now, the native function implementations are part of the move-to-yul compiler itself, so it's not something the caller can control, leading to confusion. I think we should
1. move native function implementations into move-stdlib (preferrable gated behind a feature flag)
2. define a native function interface for the move-to-yul compiler and let the caller pass implementations in, similar to how MoveVM handles its native function implementations

cc @emmazzz

wrwg replied:
I think it might be sufficient to have an error message like: "unexpected undefined native function (bug?)"? At least for now.

vgao1996 commented on language/evm/move-to-yul/src/generator.rs:
```diff
@@ -193,8 +218,29 @@ impl Generator {
             self.need_move_function(&fun_id);
+
+            for (idx, arg) in args.iter().enumerate() {
+                emit!(ctx.writer, "let arg{} := ", idx);
```
It would be great if you guys could double check if I'm using the right naming scheme here for the bindings of the arguments. I used `a0, a1, ...` initially and somehow the compiler complained about conflicting names.

wrwg replied:
Internal names should be prefixed with `$`. Source level names will be used for locals where they are available (so if you have `a0` in the source, it would potentially conflict). This kind of to make generated code better readable for debugging.

However,  `$t[0-9]+` must be avoided because the stackless bytecode uses it already for temporaries... This is not completely clean, but not that bad that refactoring was worth the effort until now.

wrwg commented on language/evm/move-to-yul/src/generator.rs:
```diff
@@ -172,17 +179,35 @@ impl Generator {
-                "test functions cannot have parameters or return values \
-                    at this point of time",
+                &format!(
+                    "test functions have {} parameters but got {} arguments",
```
test function has {} parameters but {} were provided

wrwg commented on language/tools/move-unit-test/src/test_runner.rs:
```diff
@@ -556,12 +556,30 @@ impl SharedTestingConfig {
+                None => {
+                    // Failed to generate yul code due to some user errors.
+                    // Mark test as failed.
+                    output.fail(function_name);
```
Should we perhaps add here diagnosis from the model?

vgao1996 commented:
@wrwg I made some updates:
1. Added the `$` prefix to argument bindings
2. Reverted the panic back to an error
3. Included (move-to-yul) compiler diagnostics in the test statistics so they are reported to the user
4. Other minor changes as you requested

Let me know if this looks good to you!

wrwg reviewed (approved):
Great to see this.
